### DISCUSSION TITLE
Eth Dev Assistant

### DISCUSSION THREAD

**Discussion by Spacesai1or** [ethereum/web3.js#6200] [Ideas]
### The Idea

I've started work on a POC for a replacement of `4.x`'s current testing infrastructure around creating chain instances, creating accounts, and funding accounts. In `1.x` we had a tool, [Geth Dev Assistant](https://github.com/cgewecke/geth-dev-assistant), that handled this for Geth. My idea for [Eth Dev Assistant](https://github.com/spacesailor24/eth-dev-assistant/tree/master) is to expand that offering to multiple execution/testing clients (i.e. Ganache and Hardhat) to make testing across different client software easier to manage, run locally, and debug

### The POC

I'm creating the discussion as my repo is just a POC, and now would be the time to talk architecture and scoping out the MVP. So far the architecture is pretty basic:

- [client.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/src/clients/client.ts)
    - Defines an abstract `Client` class to be used by all supported client softwares
    - Intended to use a docker image URI
    - Supports functionality such as:
        - Starting, getting the status of, and stopping docker containers
        - Creating accounts in a local web3.js wallet (with the option to pre-fund them)
        - Funding accounts using an account with pre-genesis allocated ETH
- [geth.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/src/clients/geth.ts)
    - Example implementation of a client class extending `Client` abstract class
    - Contains init code for starting docker container
- [ganache.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/src/clients/ganache.ts)
    - Example implementation of a client class extending `Client` abstract class
    - Contains init code for starting docker container
- [utils.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/src/clients/utils.ts)
    - Various util methods primarily used by `client.ts`

And that's about it for the POC. You can see in these E2E tests how the library is intended to be used:

- [client_geth.test.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/src/clients/utils.ts)
- [client_ganache.test.ts](https://github.com/spacesailor24/eth-dev-assistant/blob/master/test/e2e/client_ganache.test.ts)

### The Goal

Here is where I'd like everyone to spend time thinking about what would make this library useful. Below I will list some of the features of my POC that I deem are improvements on what `4.x` currently uses for testing:

1. Replacement of bash scripts with TS functions

Currently `4.x` uses a mixture of bash scripts and TS functions to perform operations such as starting/stopping the client software, generating and funding accounts

This is improved upon in the POC by having this functionality controllable via await-able TS functions

2.  Separate bash scripts for running each client software

Currently `4.x` has a bash script for starting/stopping Geth and a separate one for Ganache

The POC improves upon this by offering a standard interface (`client.ts`) that a client implementation class can extend, so that all client can be interacting with in a common way

3. Multiple instances of clients can be created easily

There are certain tests we have in `4.x` that step on the toes of other tests and therefore have to be ran synchronously or by generating and funding new accounts

The POC improves upon this by offering a TS function to start and stop the client software (using Docker containers), so that a global client can be created and used for tests that can be ran async, and individual clients can be started and stopped for synchronous test

_Disclaimer_ This should be possible with some modification of POC. I haven't tested this, but theoretically this should work if POC is updated to handle using different port numbers for each client instance

#### Unimplemented Features

4. Geth Dev Assistant allows for the config of the client

It should be possible to extend the POC to support client configuration using command line args when starting the Docker container

5. Support consensus clients

It should be possible to control consensus clients and setup the connection between it and an execution client for more holistic testing and for future web3.js packages such as a beaconchain package

6. Mainnet forking

It should be possible to extend the POC to support forking the mainnet chain using client software that supports forking

**Comment by luu-alex** [comment]
Thanks alot Wyatt for working on this, this looks thorough and well thought out. 
Do you think that it would be a benefit adding a server instance if available onto the client provides one? We use the ganache server instance for a [few of our e2e tests ](https://github.com/web3/web3.js/blob/031c0bfa9c722cb3d71842c4446424260a2bfc76/packages/web3-providers-ws/test/integration/ganache_fault_tolerance.test.ts#L79)

**Reply by Spacesai1or** [reply]
Not super necessary as having Ganache in a docker instance should be using the same runtime. You could just write a `Client` wrapper around the `ganache` package though if you want to cover all the bases
